Fix the issue of tf.divide's return value is not a tensor

yongtang opened the pull request:
This PR fixes the issue of #39475 where tf.divide's return value
is not a tensor in case x, y in divide(x, y) are both primitive python
types.

The reason was that tf.divide relies on implict `x / y`. However,
if both x and y are not tensor, the return value will fall through
python and will not be a tensor.

This PR fixes the issue.

This PR fixes #39475.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix the issue of tf.divide's return value is not a tensor

This PR fixes the issue of 39475 where tf.divide's return value
is not a tensor in case x, y in divide(x, y) are both primitive python
types.

The reason was that tf.divide relies on implict `x / y`. However,
if both x and y are not tensor, the return value will fall through
python and will not be a tensor.

This PR fixes the issue.

This PR fixes 39475.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

jaingaurav commented on tensorflow/python/ops/math_ops.py:
```diff
@@ -438,9 +439,13 @@ def divide(x, y, name=None):
     # override names. Use a dummy class to track the runtime division behavior
     return DivideDelegateWithName(x, name) / y
   else:
+    if not (isinstance(x, ops.Tensor)  or isinstance(y, ops.Tensor)):
```
Nit: please fix double whitespace

jaingaurav commented on tensorflow/python/ops/math_ops_test.py:
```diff
@@ -495,6 +495,12 @@ def testConsistent(self):
+  def testWithPythonValue(self):
+    # Test case for GitHub issue 39475:
+    # https://github.com/tensorflow/tensorflow/issues/39475
+    x = math_ops.divide(5,  2)
```
Nit: please fix double whitespace

jaingaurav commented on tensorflow/python/ops/math_ops_test.py:
```diff
@@ -495,6 +495,12 @@ def testConsistent(self):
+    # https://github.com/tensorflow/tensorflow/issues/39475
+    x = math_ops.divide(5,  2)
+    self.assertTrue(isinstance(x, ops.Tensor))
+
```
Nit: remove extra blank line

jaingaurav commented on tensorflow/python/ops/math_ops.py:
```diff
@@ -438,9 +439,13 @@ def divide(x, y, name=None):
+        return _truediv_python3(x, y)
     return x / y
 
-
```
Please keep blank line

jaingaurav commented on tensorflow/python/ops/math_ops.py:
```diff
@@ -438,9 +439,13 @@ def divide(x, y, name=None):
+        return _truediv_python2(x, y)
+      else:
+        return _truediv_python3(x, y)
     return x / y
```
I'm sorry, I may be missing something here, but is the logic above saying that we'll just use `_truediv_python` for non-tensor inputs? If so, then the test case shouldn't pass, since we should get a python value back.

I'm probably misreading the code. If so, please document or restructure to help clarify.

jaingaurav replied:
Okay I think I understand this PR better, I was misreading the if condition.

Wouldn't it be simpler to simply call convert to tensor on `x` and then let the operator overloads kick in? That's basically what each of those function do in the first place.

jaingaurav replied:
Please remove the comment or simply to simple say we do the convert in order always return a Tensor. Also, I think the code should be:
```
tensor_util.is_tensor(x):
  x = ops.convert_to_tensor(x)
```

jaingaurav replied:
Why do you need to check `tensor_util.is_tensor(y)`?

yongtang replied:
@jaingaurav I saw the logic in `_OverrideBinaryOperatorHelper` will try both `__truediv__` and rtruediv__` (LHS or RHS). I was overly concerned about if convert x may cause issues when y is a different dtype. So initially tried to test both.

But now I give it try with always convert x. Looks like all tests pass so it should be fine. Thanks for the help!

yongtang committed:
Explain the reason to call _truediv_python3/_div_python2 explicitly (not through registered '/'

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @jaingaurav for the review and really sorry for the confusion due to lack of comment in code.

The `tf.math.divide` will compute python style division `x / y` with output as tensor. Since '/' in tensorflow already registered for tensors (through `__div__`, `__rdiv__` in python2 and `__truediv__` and `__rtruediv__` in python 3). Eventually they are routed to `truediv_python3/_div_python2` with output as tensor.

However, '/' registered does not cover the case where both x and y are not tensors. In that situation, `truediv_python3/_div_python2`  will not be called and the python 's default '/' will take place. That means the following:
```
>>> tf.divide(5, 2)
2.5 <= should be <tf.Tensor: shape=(), dtype=float64, numpy=2.5> instead.
```

So this PR will check for input types, and, in case both x and y are not tensors, call `truediv_python3/_div_python2` to enforce the tensorflow to process `x / y`.

I have updated the PR with comments added. Please take a look and let me know if there are any issues.

yongtang committed:
Simplify comment and use tensor_util.is_tensor and convert_to_tensor if both x and y are not tensor

This is needed for `x / y` where both x and y are not tensors. As long as x is
a tensor, `x / y` will work and will rely on _truediv_python3 to correctly
figure out the right type for y.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @jaingaurav. That makes sense. I have updated the PR now. Please take a look.

yongtang commented:
Thanks @jaingaurav for the review. The PR has been updated. Please take a look.

yongtang committed:
Remove the need to check if y is a tensor (always convert x if not)

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@jaingaurav The PR has been updated now. Please take a look and let me know if there are any other issues.

jaingaurav reviewed (approved):
Looks good. Note that I might have to run some exhaustive tests internally since this does change the API behaviour in a subtle manner.

yongtang commented:
@jaingaurav Thanks! Let me know if additional update is needed to pass the internal tests.
